cips/cip-019.md

```diff
@@ -313,6 +314,99 @@ Namespace data may span over multiple rows, in which case all the data is encaps
+
+[**EdsID**](#edsid): A EdsID of the NamespaceData container. It MUST follow [EdsID](#edsid) formatting and field validity rules.
+
+[**Namespace**][ns]: A fixed-size 29 bytes array representing the Namespace of interest. It MUST follow [Namespace][ns]
+formatting and its validity rules.
+
+[Serialized](#serialization-for-share-identifiers) NamespaceDataID MUST have a length of 37 bytes.
+
+#### NamespaceData Container
+
+NamespaceData containers encapsulate user-submitted data under [namespaces][ns] within an entire [DataSquare][square]. Namespace data may span over multiple rows.
+
+NamespaceData containers are protobuf formatted using the following proto3 schema:
+
+```protobuf
+syntax = "proto3";
+
+message NamespaceData {
+  repeated RowNamespaceData namespaceData = 1;
+}
+```
+
+The fields with validity rules that form Data containers are:
+
+**NamespaceData**: A variable size [RowNamespaceData](#rownamespacedata-container) array representing data of a namespace in a [DataSquare][square].
+
+#### RangeNamespaceDataID
+
+RangeNamespaceDataID identifies a continuous range of shares under the same namespace within an ODS.
+
+RangeNamespaceDataID identifiers formatted as shown below:
+
+```text
+RangeNamespaceDataID {
+    EdsID;
+    From: u16;
+    To: u16;
+}
+```
+
+The fields with validity rules that form RangeNamespaceDataID are:
+
+[**EdsID**](#edsid): A EdsID of the RangeNamespaceData container. It MUST follow [EdsID](#edsid) formatting and field validity rules.
+
+**From**: A uint16 representing the first index of the sampled share within the ODS. The 16 bit
+limit fits data squares up to 2TB. It MUST stay within the number of original shares(1st quadrant) of the EDS.
+
+**To**: A uint16 representing the last index of the sampled share within the ODS(exclusively). The 16 bit
+limit fits data squares up to 2TB. It MUST stay within the number of original shares(1st quadrant) increased by one of the EDS.
+
+#### RangeNamespaceData Container
+
+RangeNamespaceData containers encapsulate user-submitted data under speicific coordinates [from;to] within a [DataSquare][square].
+
+RangeNamespaceData containers are protobuf formatted using the following proto3 schema:
+
+```protobuf
+syntax = "proto3";
+
+message RangeNamespaceData {
+  repeated RowShares shares = 1;
+  Proof firstIncompleteRowProof = 2;
+  Proof lastIncompleteRowProof = 3;
+}
+```
+
+The fields with validity rules that form Data containers are:
+
+**Shares**: A variable size [RowShares](#rowshares) array representing data of a range in a single [DataSquare][square].
+
+**FirstIncompleteRowProof**: An [NMT][nmt] proof for the first row, present when:
+
+- The range starts at a non-zero share index, OR
+- In single-row ranges, when the range is not full (less than the number of shares in the row)
+
+**LastIncompleteRowProof**: An [NMT][nmt] proof for the last row in multi-row ranges, present when the range ends before the row's final share.
+
+Both Proofs can be empty.
```

Comment:
how can someone prove the data? `RowShares` don't carry any proofs.
When having only the ODS part of the row, the only thing that comes to my mind is that one needs to recompute the parity data, create an Nmt of the whole row and then compare the root with the DAH? Sounds like quite a lot of computation, especially for bigger squares